### DISCUSSION TITLE
Poetry + Private Gitlab Repository - Not working

### DISCUSSION THREAD

**Discussion by devandreassimmeth** [python-poetry/poetry#7083] [Q&A] [answered]
Hi all together,

I have a private Gitlab Project with a Package published in the Package Registry. Now I try to use this in a different project using poetry as dependency manager. I tried it as described in this [issue comment](https://github.com/python-poetry/poetry/issues/2062#issuecomment-1056793536).

So the packaging in the CI pipline went well and the package is available in the package registry. To do this, I followed this [tutorial](https://docs.mpcdf.mpg.de/doc/data/gitlab/devop-tutorial.html).

Then I created a deploy token with scope `read_package_registry` and added the source to my "destination" project like in the issue comment described. But every time I try to add the project over `project add ...` I get the following error:

`401 Client Error: Unauthorized for url: https://gitlab.com/api/v4/projects/<projectId>/packages/pypi/files/<someNumber>/<project-name>-0.0.1-py3-none-any.whl#sha256=<sha256Identifier>`

When I open the url in the browser, the .whl file is downloaded without any unauthorized error. And also with pip and pipenv the installation of the package works with the same URL as in the issue comment described. So it's not a error in the package or GIT itself.

I searched a lot, but I can't find any information why I have this problem with poetry. Has anyone of you an idea, what is wrong here?

I'm using the following versions: 
- poetry: 1.2.2
- poetry-core: 1.4.0 (without specifier in "build-system/requires")

Thanks for any answer/help!

**Comment by squirrel532** [comment, marked as answer]
You can try the solution in #6799

**Reply by devandreassimmeth** [reply]
Nice, that also worked for me. Thx :)

**Reply by UncleGoogle** [reply]
which one? Putting creds in poetry.toml does not solves the problem

**Comment by DonXavierdev** [comment]
It worked for me when i specifically added the url as a command 
```
poetry add "py-auth@https://__token__:<your-token>@gitlab.com/api/v4/projects/61156295/packages/pypi/files/0352517b78320700a2ebd529462c7cee4a864f71e99d009b140dffa0de12aac5/py_auth-0.1.0-py3-none-any.whl"
```

**Comment by sebhub** [comment]
The information here and in #6799 was very helpful to use poetry 2.0.1 in our private Gitlab instance.

# How to use the Gitlab package registry on a development host

You need a [personal access token](https://docs.gitlab.com/ee/user/packages/package_registry/).

Add the following section to `~/.netrc`:

```
machine <gitlab-host>
login __token__
password <your-personal-access-token>
```

If you want to install packages from this registry on a development host through ``pip`` use:

```
pip install --index-url https://<gitlab-host>/gitlab/api/v4/projects/<project-id>/packages/pypi/simple --cert=<your.pem> <package-name>
```

If you want to add packages from this registry on a development host through ``poetry`` use:

```
poetry source add --priority=supplemental gitlab https://<gitlab-host>/gitlab/api/v4/projects/<project-id>/packages/pypi/simple
poetry config repositories.gitlab https://<gitlab-host>/gitlab/api/v4/projects/<project-id>/packages/pypi
poetry config certificates.gitlab.cert <your.pem>
poetry add --source=gitlab <package-name>
```

# How to use the Gitlab package registry in a CI job

You need the ``$CI_JOB_TOKEN``.  Make sure the project associated with the CI job is in the CI/CD job token allow list.

Use the following job script template to publish the package using ``poetry``:

```yaml
script:
  - echo -e "machine <gitlab-host>\nlogin gitlab-ci-token\npassword ${CI_JOB_TOKEN}\n" >> ~/.netrc
  - poetry config repositories.gitlab ${CI_API_V4_URL}/projects/<project-id>/packages/pypi
  - poetry config certificates.gitlab.cert false
  - poetry install
  - make
  - poetry build
  - poetry publish --repository gitlab
```

Use the following job script template to install the package using ``pip``.

```yaml
script:
  - echo -e "machine git\nlogin gitlab-ci-token\npassword ${CI_JOB_TOKEN}\n" >> ~/.netrc
  - pip3 install --index-url ${CI_API_V4_URL}/projects/<project-id>/packages/pypi/simple --trusted-host <gitlab-host> <package-name>
```

**Reply by NaelsonDouglas** [reply]
the ~/.netrc solution worked for me. Thanks!

**Comment by fruitl00p** [comment]
@sebhub i like the .netrc solution as it more closely mimics the local dev environment. However i'm struggling with versionable (released) dependencies using poetry. Should that work the same way?
